content/blog/fast-data-processing-pipeline-for-predicting-flight-delays-using-apache-apis-kafka-spark-streaming-and-machine-learning-part-2.md

```diff
@@ -0,0 +1,231 @@
+
+![Kafka API](/img/kafka-api.png)
+
+You can think of a partition like a queue; new messages are appended to the end, and messages are delivered in the order they are received.
+
+![MapR Cluster](/img/mapr-cluster.png)
+
+![Messages](/img/messages.png)
+
+Unlike a queue, messages are not deleted when read; they remain on the partition, available to other consumers. Messages, once published, are immutable, and can be retained forever. 
+
+![Not deleting messages](/img/not-deleting-messages.png)
+
+Not deleting messages when they are read allows for high performance at scale and also for processing of the same messages by different consumers for different purposes such as multiple views with polyglot persistence.
+
+![New subscribers of information](/img/new-subscribers-information.png)
+
+New subscribers of information can replay the data stream, specifying a starting point as far back as the data retention policy enables.
+
+![Create new view, index, cache](/img/create-new.png)
+
+![Read From new View](/img/read-from-new-view.png)
+
+## Data Pipelines
+
+When you combine these messaging capabilities with the simple concept of microservices, you can greatly enhance the agility with which you build, deploy, and maintain complex data pipelines. Pipelines are constructed by simply chaining together multiple microservices, each of which listens for the arrival of some data, performs its designated task, and optionally publishes its own messages to a topic. Development teams can deploy new services or service upgrades more frequently and with less risk, because the production version does not need to be taken offline. Both versions of the service simply run in parallel, consuming new data as it arrives and producing multiple versions of output. Both output streams can be monitored over time; the older version can be decommissioned when it ceases to be useful.
+
+## Machine Learning Logistics and Data Pipelines
+
+Combining data pipelines with machine learning can handle the logistics of machine learning in a flexible way by:
+
+* Making input and output data available to independent consumers
+* Managing and evaluating multiple models and easily deploying new models
+
+![Machine Learning Logistics and Data Pipelines](/img/machine-learning-logistics.png)
+
+Architectures for these types of applications are discussed in more detail in the ebooks Machine Learning logistics, Streaming Architecture, and Microservices and Containers.
```

Comment:
Were these ebooks supposed to have been hotlinked?

![image](https://user-images.githubusercontent.com/77496261/137188039-29d69832-6d2b-4477-b332-0b4016d5bcaa.png)

Reply:
Added.